scapy/contrib/pnio_dcp.py

```diff
@@ -0,0 +1,559 @@
+# coding: utf8
+
+# This file is part of Scapy
+# Scapy is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# any later version.
+#
+# Scapy is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Scapy. If not, see <http://www.gnu.org/licenses/>.
+
+# Copyright (C) 2019 Stefan Mehner (stefan.mehner@b-tu.de)
+
+# scapy.contrib.description = Profinet DCP layer
+# scapy.contrib.status = loads
+
+from __future__ import absolute_import
+from scapy.all import Packet, bind_layers, Ether, UDP, Padding
+from scapy.fields import ByteEnumField, ShortField, XShortField, ShortEnumField, FieldLenField, XByteField, ByteField, \
+    XIntField, MultiEnumField, IPField, MACField, StrLenField, PacketListField, PadField, ConditionalField, LenField
+
+
+# minimum packet is 60 bytes.. 14 bytes are Ether()
+MIN_PACKET_LENGTH = 44
+
+#####################################################
+#                     Constants                     #
+#####################################################
+
+DCP_GET_SET_FRAME_ID = 0xFEFD
+DCP_IDENTIFY_REQUEST_FRAME_ID = 0xFEFE
+DCP_IDENTIFY_RESPONSE_FRAME_ID = 0xFEFF
+
+DCP_REQUEST = 0
+DCP_RESPONSE = 1
+
+DCP_SERVICE_ID_GET = 0x03
+DCP_SERVICE_ID_SET = 0x04
+DCP_SERVICE_ID_IDENTIFY = 0x05
+
+DCP_SERVICE_ID = {
+    0x00: "reserved",
+    0x01: "Manufacturer specific",
+    0x02: "Manufacturer specific",
+    0x03: "Get",
+    0x04: "Set",
+    0x05: "Identify",
+    0x06: "Hello",
+}
+
+DCP_SERVICE_TYPE = {
+    0x00: "Request",
+    0x01: "Response Success",
+    0x05: "Response - Request not supported",
+}
+
+DCP_DEVICE_ROLES = {
+    0x00: "IO Supervisor",
+    0x01: "IO Device",
+    0x02: "IO Controller",
+
+}
+
+DCP_OPTIONS = {
+    0x00: "reserved",
+    0x01: "IP",
+    0x02: "Device properties",
+    0x03: "DHCP",
+    0x04: "Reserved",
+    0x05: "Control",
+    0x06: "Device Initiative",
+    range(0x07, 0x7f): "reserved",
```

Comment:
This won't work in Python.

We typically have `YesNoByteField` (yes we need a better name :smile:) for that purpose, see

`help(YesNoByteField)`

Reply:
What exactly do you mean here? The enums as dicts and the ranges within should work. 

At what point do you suggest the YesNoByteField? If you mean it for _DCP_Request_ and _DCP_Response_ constants in my opnion it is not the right thing, since it is a service type for the frame and not a boolean expression.

Reply:
oh a range within a key in a dictionary does not work. Is it okay to implement it in that way:


```python
reserved = { i: "reserved" for i in range(0x0003, 0x007f)}
IP_BLOCK_INFOS = {
    0x0000: "IP not set",
    0x0001: "IP set",
    0x0002: "IP set by DHCP",
    # range(0x0003, 0x007f): "reserved",
    0x0080: "IP not set (address conflict detected)",
    0x0081: "IP set (address conflict detected)",
    0x0082: "IP set by DHCP (address conflict detected)",
}
IP_BLOCK_INFOS.update(reserved)
```

Reply:
That would be OK to me.